web/src/echo.ts

```diff
@@ -436,6 +438,28 @@ export function edit_locally(message: Message, request: LocalEditRequest): Messa
     return message;
 }
 
+export function update_topic_hash_to_contain_with_term(message: Message): void {
+    // If the current filter consists only of channel and topic terms
+    // and the incoming message belongs to same narrow, we try to
+    // update the hash, to convert to permalink.
+    if (message.type !== "stream") {
+        return;
+    }
+    const filter = message_lists.current!.data.filter;
+    if (
+        filter.is_in_channel_topic_narrow() &&
+        filter.has_operand("channel", message.stream_id.toString()) &&
+        filter.has_operand("topic", message.topic)
+    ) {
+        const narrow_terms = filter.terms();
+        const with_term = {operator: "with", operand: message.id.toString()};
+
+        narrow_terms.push(with_term);
+        const new_hash = hash_util.search_terms_to_hash(narrow_terms);
+        browser_history.set_hash(new_hash);
+    }
+}
```

Comment:
This seems like it should be a separate commit, as it's a separable feature.

Reply:
Oh, is this #31152 being in this PR too?

Reply:
OK, I'm going to drop this from this PR, since I think it's the most risky part, and doing so might let me get a lot closer to merging the rest.